sphereon-kmp-crypto-kms-azure/src/jvmMain/kotlin/com/sphereon/crypto/kms/azure/AzureKeyVaultCryptoProvider.jvm.kt

```diff
@@ -0,0 +1,389 @@
+            .setHeaders(headers?.map { com.azure.core.util.Header(it.name, it.values) })
+    }
+
+    private fun ExponentialBackoffRetryOpts.toExponentialBackoffOptions(): ExponentialBackoffOptions {
+        return ExponentialBackoffOptions()
+            .setMaxRetries(maxRetries)
+            .setBaseDelay(if (baseDelayInMS == null) null else Duration.ofMillis(baseDelayInMS))
+            .setMaxDelay(if (maxDelayInMS == null) null else Duration.ofMillis(maxDelayInMS))
+    }
+
+    private fun CredentialOpts.toTokenCredential(tenantId: String): TokenCredential {
+        return when (credentialMode) {
+            CredentialMode.SERVICE_CLIENT_SECRET -> secretCredentialOpts?.toClientSecretCredential(tenantId)
+                ?: throw SignClientException("No client secret options provided")
+
+            CredentialMode.SERVICE_CLIENT_CERTIFICATE -> certificateCredentialOpts?.toClientCertificateCredential(
+                tenantId
+            )
+                ?: throw SignClientException("No client certificate options provided")
+
+            CredentialMode.USER_INTERACTIVE_BROWSER -> interactiveBrowserCredentialOpts?.toInteractiveBrowserCredential(
+                tenantId
+            )
+                ?: throw SignClientException("No interactive browser options provided")
+
+            CredentialMode.USER_USERNAME_PASSWORD -> usernamePasswordCredentialOpts?.toUsernamePasswordCredential(
+                tenantId
+            )
+                ?: throw SignClientException("No username password options provided")
+        }
+    }
+
+    private fun SecretCredentialOpts.toClientSecretCredential(tenantId: String): ClientSecretCredential {
+        return ClientSecretCredentialBuilder()
+            .clientId(clientId)
+            .clientSecret(clientSecret)
+            .tenantId(tenantId)
+            .build()
+    }
+
+    private fun CertificateCredentialOpts.toClientCertificateCredential(tenantId: String): ClientCertificateCredential {
+        return ClientCertificateCredentialBuilder()
+            .clientId(clientId)
+            .pemCertificate(pemCertificatePath)
+            .tenantId(tenantId)
+            .build()
+    }
+
+    private fun UsernamePasswordCredentialOpts.toUsernamePasswordCredential(tenantId: String): UsernamePasswordCredential {
+        return UsernamePasswordCredentialBuilder()
+            .clientId(clientId)
+            .username(userName)
+            .password(password)
+            .tenantId(tenantId)
+            .build()
+    }
+
+    private fun InteractiveBrowserCredentialOpts.toInteractiveBrowserCredential(tenantId: String): InteractiveBrowserCredential {
+        return InteractiveBrowserCredentialBuilder()
+            .clientId(clientId)
+            .redirectUrl(redirectUrl)
+            .tenantId(tenantId)
+            .build()
+    }
+
+    private fun ByteArray.toBase64UrlString(): String =
+        java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(this)
```

Comment:
We already have base64 url encoding/decoding in the KMP lib. No need to use Java for it